ci: nested SDS E2E on skeleton

yachmenevas opened the pull request:
Description

Add a minimal nested SDS E2E on top of the skeleton.
Preserves triggers, dependencies, and ordering; replaces Ceph-only parts with SDS enablement (source: deckhouse) and nested replicated storage setup.

Why do we need it, and what problem does it solve?

Validate triggers/dependencies and SDS profile flow inside nested clusters before introducing full tests/notifications.

What is the expected result?

Workflow runs on schedule/dispatch/PR, boots a nested cluster, enables SDS modules, creates LVMVolumeGroups + ReplicatedStoragePool + ReplicatedStorageClass, and proceeds to E2E gates.

Checklist

- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [ ] Changes were tested in the Kubernetes cluster manually (workflow_dispatch).

Changelog entries

section: ci
type: chore
summary: "Add nested SDS E2E on top of skeleton (deckhouse source)."

Isteb4k commented on ci/dvp-e2e/charts/infra/templates/ingress.yaml:
```diff
@@ -0,0 +1,74 @@
+  namespace: {{ .Values.namespace }}
+spec:
+  ports:
+  - port: 80
```
Do we really need both? Both 80 and 443?

yachmenevas replied:
Kept HTTPS only (443)

yachmenevas committed:
ci: nested SDS setup only (prepare+cleanup); remove E2E tests and report; restore jump-host; stabilize SDS bring-up (MPO + explicit ModuleConfigs + CRD‑compatible LVG + safe default SC patch)

Signed-off-by: Anton Yachmenev <anton.yachmenev@flant.com>

Isteb4k commented on ci/dvp-e2e/values.yaml:
```diff
@@ -0,0 +1,69 @@
+    virtualDisks:
+      os: ceph-pool-r2-csi-rbd
+      data: ceph-pool-r2-csi-rbd
+  sds-local:
```
“We’re not testing local and ceph right now. Let’s remove them from the configuration for now. Everything necessary for them will be added in a separate pr

yachmenevas replied:
Done

Isteb4k commented on .github/workflows/e2e-matrix.yml:
```diff
@@ -169,12 +198,154 @@ jobs:
             RUN_DIR="${{ env.TMP_ROOT }}/runs/${{ env.RUN_ID }}"
           echo "VALUES_TEMPLATE_FILE=${{ env.TMP_ROOT }}/runs/${{ env.RUN_ID }}/values.yaml" >> $GITHUB_ENV
 
+      - name: Configure registry auth (REGISTRY_DOCKER_CFG)
```
REGISTRY_DOCKER_CFG was renamed

Isteb4k commented on .github/workflows/e2e-matrix.yml:
```diff
@@ -169,12 +198,154 @@ jobs:
+        run: |
+          prod_user="${{ secrets.PROD_READ_REGISTRY_USER }}"
+          prod_pass="${{ secrets.PROD_READ_REGISTRY_PASSWORD }}"
+          dev_user="${{ secrets.BOOTSTRAP_DEV_REGISTRY_LOGIN }}"
```
BOOTSTRAP_DEV_REGISTRY_LOGIN was removed

Isteb4k commented on .github/workflows/e2e-matrix.yml:
```diff
@@ -169,12 +198,154 @@ jobs:
+          echo "SNAPSHOT_STORAGE_CLASS=${SNAPSHOT_STORAGE_CLASS}" >> $GITHUB_ENV
+          echo "ATTACH_DISK_SIZE=${ATTACH_DISK_SIZE}" >> $GITHUB_ENV
+          echo "storage_class=$STORAGE_CLASS" >> $GITHUB_OUTPUT
+          echo "image_storage_class=$IMAGE_STORAGE_CLASS" >> $GITHUB_OUTPUT
```
Why in lowercase?

Isteb4k commented on .github/workflows/e2e-matrix.yml:
```diff
@@ -169,12 +198,154 @@ jobs:
+          echo "snapshot_storage_class=$SNAPSHOT_STORAGE_CLASS" >> $GITHUB_OUTPUT
+          echo "attach_disk_size=$ATTACH_DISK_SIZE" >> $GITHUB_OUTPUT
+          # Pass storage profile into run values for Helm templates
           PROFILE='sds' yq eval --inplace '.storageProfile = strenv(PROFILE)' "${{ env.TMP_ROOT }}/runs/${{ env.RUN_ID }}/values.yaml"
```
Why are you overriding the PROFILE env?

Isteb4k commented on .github/workflows/e2e-matrix.yml:
```diff
@@ -169,12 +198,154 @@ jobs:
+            NESTED_DIR="${{ env.TMP_ROOT }}/runs/${{ env.RUN_ID }}/nested" \
+            NESTED_KUBECONFIG="${{ env.TMP_ROOT }}/runs/${{ env.RUN_ID }}/nested/kubeconfig"
+
+      # Ingress smoke disabled: not required for storage config
```
Do we need these comments?

yachmenevas replied:
Deleted

Isteb4k commented on .github/workflows/e2e-matrix.yml:
```diff
@@ -225,7 +372,7 @@ jobs:
-          version: 'latest'
+          version: "latest"
 
       - name: Build parent kubeconfig from secret (cleanup)
```
The build configuration is duplicated here. Let’s move it into a .sh script.

yachmenevas replied:
Done

yachmenevas commented on ci/dvp-e2e/Taskfile.yaml:
```diff
@@ -0,0 +1,1097 @@
+        echo "[INFRA] All worker VMs configured with storage disks via hotplug"
+
+  infra:attach-worker-disks:
+    desc: Attach additional data disks to worker VMs (for SDS)
```
Only for this PR

yachmenevas committed:
ci(e2e): clean values and debug; simplify SDS wiring; minimal helper scripts; HTTPS-only ingress; param cleanup wait\n\n- remove report scripts and local tests task\n- drop unused values (storageProfiles, infra/virtualDisks/security, extra features)\n- add build_parent_kubeconfig.sh + inject_registry_cfg.sh and use them in workflow\n- call task nested:storage:sds directly from workflow; remove wrapper\n- trim logs (no lsblk, no set -x); keep concise status\n- ingress: drop 80/wildcard-http; keep HTTPS passthrough only\n- cleanup: param prefix + wait for namespace deletion

Signed-off-by: Anton Yachmenev <anton.yachmenev@flant.com>

yachmenevas committed:
ci(e2e): workflow fixes and cleanups for SDS-only PR

- use DEV_REGISTRY_DOCKER_CFG only; inject into values; login via docker cfg
- fix GitHub Actions 'if' expressions (no secrets.* in if; use env)
- add Apache-2.0 headers to helper scripts (dmtlint)
- format ci/dvp-e2e/Taskfile.yaml with repo Prettier (prettier)
- trim push triggers to ci-e2e-nested-sds only; remove noisy comments
- keep setup/prepare/cleanup only (tests/report moved to next PR)

Signed-off-by: Anton Yachmenev <anton.yachmenev@flant.com>

yachmenevas committed:
ci(e2e): workflow fixes and cleanups for SDS-only PR\n\n- top-level concurrency + unified job concurrency (head_ref||ref_name) to prevent duplicate runs\n- use DEV_REGISTRY_DOCKER_CFG only; inject into values; login via docker cfg\n- fix Actions 'if' expressions (no secrets.* in if; use env)\n- add Apache-2.0 headers to helper scripts (dmtlint)\n- format ci/dvp-e2e/Taskfile.yaml with repo Prettier (prettier)\n- trim push triggers to ci-e2e-nested-sds only; remove noisy comments

Signed-off-by: Anton Yachmenev <anton.yachmenev@flant.com>

yachmenevas commented on ci/dvp-e2e/charts/cluster-config/values.yaml:
```diff
@@ -0,0 +1,77 @@
+  workers:
+    root: ceph-pool-r2-csi-rbd-immediate
+
+# Infrastructure components
```
Fixed

yachmenevas committed:
ci(e2e): remove unused task infra:attach-worker-disks (redundant; replaced by infra:attach-storage-disks-hotplug)

Signed-off-by: Anton Yachmenev <anton.yachmenev@flant.com>

yachmenevas committed:
ci(e2e): hardcode kubernetesVersion=Automatic in ClusterConfiguration and drop deckhouse.kubernetesVersion from values

Signed-off-by: Anton Yachmenev <anton.yachmenev@flant.com>

yachmenevas commented on ci/dvp-e2e/charts/cluster-config/templates/cluster-config.yaml:
```diff
@@ -0,0 +1,48 @@
+  prefix: {{ .Values.clusterConfigurationPrefix | default "demo-cluster" }}
+podSubnetCIDR: 10.112.0.0/16
+serviceSubnetCIDR: 10.223.0.0/16
+kubernetesVersion: "{{ .Values.deckhouse.kubernetesVersion }}"
```
Done

yachmenevas commented on ci/dvp-e2e/Taskfile.yaml:
```diff
@@ -0,0 +1,1097 @@
+
+        echo "[INFRA] All worker VMs configured with storage disks via hotplug"
+
+  infra:attach-worker-disks:
```
Deleted

yachmenevas commented on ci/dvp-e2e/Taskfile.yaml:
```diff
@@ -0,0 +1,1097 @@
+          sleep 10
+        done
+
+  nested:storage:configure:
```
Fixed

yachmenevas committed:
ci(e2e): move namespace cleanup logic into Taskfile (task cleanup:namespaces) and call it from workflow; prefix stays configurable

Signed-off-by: Anton Yachmenev <anton.yachmenev@flant.com>

yachmenevas committed:
ci(e2e): address review — move kubeconfig/registry steps into Taskfile (parent:kubeconfig, values:inject-registry); workflow calls tasks; keep params minimal

Signed-off-by: Anton Yachmenev <anton.yachmenev@flant.com>

yachmenevas committed:
ci(e2e): fix infra:attach-storage-disks-hotplug stray duplicate loop (caused unmatched 'done'); ensure Task installed before calling tasks in cleanup

Signed-off-by: Anton Yachmenev <anton.yachmenev@flant.com>

yachmenevas commented on .github/workflows/e2e-matrix.yml:
```diff
@@ -169,12 +198,154 @@ jobs:
+            VALUES_FILE="${{ env.TMP_ROOT }}/runs/${{ env.RUN_ID }}/values.yaml" \
+            PARENT_KUBECONFIG="${KUBECONFIG}" \
+            SSH_FILE_NAME="id_ed" \
+            TARGET_STORAGE_CLASS="ceph-pool-r2-csi-rbd-immediate"
```
Fixed - using dynamic $PROFILE

yachmenevas committed:
ci(e2e): address PR review comments

- Add parent_storage_class to profiles.json for dynamic SC selection
- Use profile-based PARENT_STORAGE_CLASS instead of hardcoded ceph SC
- Simplify storageClasses config to single storageClass value
- Remove PROFILE override, use dynamic profile variable
- Update get_profile_config.sh to export PARENT_STORAGE_CLASS
- Remove unused second parameter from get_profile_config.sh
- Clarify environment/outputs export comment
- Use full profile name 'sds-replicated-volume' everywhere

Isteb4k commented on .github/workflows/e2e-matrix.yml:
```diff
@@ -39,49 +35,77 @@ on:
+          cd ci/dvp-e2e
+          # Map sds-replicated-volume to sds profile from profiles.json
+          PROFILE=$(jq -r '.[0].name' profiles.json)
+          echo "profile=$PROFILE" >> "$GITHUB_OUTPUT"
```
The profile hasn’t been added to the output for this job. Will it really work like this?

yachmenevas replied:
Fixed - added profile output

Isteb4k commented on .github/workflows/e2e-matrix.yml:
```diff
@@ -39,49 +35,77 @@ on:
+      storage_class: ${{ steps.profile-config.outputs.storage_class }}
+      image_storage_class: ${{ steps.profile-config.outputs.image_storage_class }}
+      snapshot_storage_class: ${{ steps.profile-config.outputs.snapshot_storage_class }}
+      attach_disk_size: ${{ steps.profile-config.outputs.attach_disk_size }}
```
Why is this in the outputs? Remove from the outputs everything you’re not using in this PR.

Isteb4k commented on .github/workflows/e2e-matrix.yml:
```diff
@@ -169,12 +179,135 @@ jobs:
+          dev_cfg_b64='${{ secrets.DEV_REGISTRY_DOCKER_CFG }}'
+          if [ -n "$dev_cfg_b64" ]; then
+            echo "::add-mask::$dev_cfg_b64"
+            echo "REGISTRY_DOCKER_CFG=$dev_cfg_b64" >> "$GITHUB_ENV"
```
Why do you need a separate REGISTRY_DOCKER_CFG variable? Why not just always use secrets.DEV_REGISTRY_DOCKER_CFG?

yachmenevas replied:
Fixed

Isteb4k commented on .github/workflows/e2e-matrix.yml:
```diff
@@ -169,12 +179,135 @@ jobs:
+          echo "Attach Disk Size: ${ATTACH_DISK_SIZE}"
+
+          # Export variables to GitHub Actions environment and outputs
+          echo "STORAGE_CLASS=${STORAGE_CLASS}" >> $GITHUB_ENV
```
Let’s remove everything that isn’t being used right now. Only keep the jobs for deploying and deleting clusters according to the profiles in profiles.json

Isteb4k commented on .github/workflows/e2e-matrix.yml:
```diff
@@ -169,12 +179,135 @@ jobs:
+      - name: Install infra (namespace/RBAC/ingress)
+        working-directory: ci/dvp-e2e
+        run: |
+          USE_GH_SSH_KEYS=true SSH_FILE_NAME=id_ed task render-infra \
```
In CI, call a single task — task install:nested:env (or something like that). This task will then call the others.
Like this:
```
```yaml
 install-nested-env:
    cmds:
    - task: render-infra
    - task: infra-deploy
    - task: dhctl-bootstrap
    - task: infra:attach-storage-disks-hotplug
```
```

Isteb4k commented on .github/workflows/e2e-matrix.yml:
```diff
@@ -185,101 +318,47 @@ jobs:
+          echo "KUBECONFIG=$KCFG" >> "$GITHUB_ENV"
+
+      - name: Install Task
+        uses: arduino/setup-task@v2
```
You already installed it a bit earlier

yachmenevas replied:
Removed duplicate

Isteb4k commented on ci/dvp-e2e/Taskfile.yaml:
```diff
@@ -0,0 +1,980 @@
+    vars:
+      TMP_DIR: '{{ .TMP_DIR | default (printf "%s/%s" .TMP_ROOT "default") }}'
+      SSH_DIR: '{{ .SSH_DIR | default (printf "%s/%s" .TMP_DIR "ssh") }}'
+      SSH_PRIV_KEY_FILE: '{{ printf "%s/%s" .SSH_DIR .SSH_FILE_NAME }}'
```
You’re passing the global variable SSH_FILE_NAME with the value cloud through a long chain of nesting. Why not just use this global variable here directly?

yachmenevas committed:
Update ci/dvp-e2e/charts/cluster-config/values.yaml

Co-authored-by: Nikita Korolev <141920865+universal-itengineer@users.noreply.github.com>
Signed-off-by: Anton Yacmenev <anton.yachmenev@flant.com>

yachmenevas committed:
chore(ci): refactor E2E workflow and Taskfile

- Remove duplicates: kubeconfig in cleanup, run_id outputs, Install Task
- Remove hardcoded storage class values, use profile variables
- Simplify REGISTRY_DOCKER_CFG usage (direct secret access)
- Remove USE_GH_SSH_KEYS variable (always import from GitHub)
- Remove fetch-depth: 0 (unused)
- Remove unused nfs config from cluster-config/values
- Add profile to setup-nested-envs outputs
- Replace d8 installation with werf/trdl/actions/setup-app@v0.12.2
- Remove PASSWORD_FILE variable (local only)
- Change default SSH_FILE_NAME to id_ed
- Create unified task install:nested:env for complete setup
- Add Docker registry auth via config.json
- Ensure nested dir exists before creating kubeconfig

universal-itengineer commented on ci/dvp-e2e/scripts/configure_sds_storage.sh:
```diff
@@ -0,0 +1,259 @@
+  imageTag: main
+  scanInterval: 15s
+EOF
+
```
I believe that we need to wait if the module is ready
Use `kubectl wait module sds-node-configurator --for jsonpath='{.status.phase}'=Ready --timeout=60s`

universal-itengineer commented on ci/dvp-e2e/scripts/configure_sds_storage.sh:
```diff
@@ -0,0 +1,284 @@
+        - name: test-volume
+          mountPath: /mnt/disk
+  volumes:
+    - name: test-volume
```
Same as https://github.com/deckhouse/virtualization/pull/1693#discussion_r2536930928

yachmenevas replied:
Removed

universal-itengineer commented on ci/dvp-e2e/scripts/build_parent_kubeconfig.sh:
```diff
@@ -0,0 +1,60 @@
+
+# Usage:
+#   build_parent_kubeconfig.sh -o /path/to/kubeconfig -a https://api.server -t <token>
+
```
Where is this script used?

universal-itengineer commented on .github/workflows/e2e-matrix.yml:
```diff
@@ -169,12 +170,69 @@ jobs:
+          VALS="${{ env.TMP_ROOT }}/runs/${{ env.RUN_ID }}/values.yaml"
+          task values:inject-registry VALUES_FILE="$VALS" REGISTRY_DOCKER_CFG='${{ secrets.DEV_REGISTRY_DOCKER_CFG }}'
+
+      - name: Configure Docker registry auth
```
Why we need this step?
To deploy a cluster, it is enough to use `secrets.DEV_REGISTRY_DOCKER_CFG`

universal-itengineer commented on .github/workflows/e2e-matrix.yml:
```diff
@@ -169,12 +170,69 @@ jobs:
+          mkdir -p ~/.docker
+          echo '${{ secrets.DEV_REGISTRY_DOCKER_CFG }}' | base64 -d > ~/.docker/config.json
+
       - name: Configure storage profile
```
I believe that at this stage it is sufficient to use only the `jq` utility to define the storage profile from json file.

yachmenevas commented on .github/workflows/e2e-matrix.yml:
```diff
@@ -39,49 +35,77 @@ on:
+          echo "run_id=$RUN_ID" >> "$GITHUB_OUTPUT"
+          mkdir -p ./tmp/run-context
+          echo "profile: ${PROFILE}" > ./tmp/run-context/config.yaml
+          echo "run_id: ${RUN_ID}" >> ./tmp/run-context/config.yaml
```
Fixed - removed duplicate run_id from prepare job

yachmenevas commented on ci/dvp-e2e/Taskfile.yaml:
```diff
@@ -0,0 +1,980 @@
+    vars:
+      TMP_DIR: '{{ .TMP_DIR | default (printf "%s/%s" .TMP_ROOT "default") }}'
+      PASSWORD_FILE: '{{ printf "%s/%s" .TMP_DIR "password.txt" }}'
+      PASSWORD_HASH_FILE: '{{ printf "%s/%s" .TMP_DIR "password-hash.txt" }}'
```
Fixed - now defined once per task

yachmenevas commented on ci/dvp-e2e/scripts/inject_registry_cfg.sh:
```diff
@@ -0,0 +1,38 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+set -euo pipefail
+
```
Removed

yachmenevas commented on ci/dvp-e2e/scripts/configure_sds_storage.sh:
```diff
@@ -0,0 +1,259 @@
+  fi
+  echo "[SDS] sds-replicated-volume status: ${STATUS:-NotFound}, retry $i/60"
+  sleep 10
+done
```
Fixed

Isteb4k commented on .github/workflows/e2e-matrix.yml:
```diff
@@ -167,119 +148,94 @@ jobs:
+          echo "PARENT_STORAGE_CLASS=${PARENT_STORAGE_CLASS}" >> $GITHUB_ENV
+          echo "ATTACH_DISK_SIZE=${ATTACH_DISK_SIZE}" >> $GITHUB_ENV
+          # Pass storage profile into run values for Helm templates
+          yq eval --inplace ".storageProfile = \"${PROFILE}\"" "${{ env.TMP_ROOT }}/runs/${{ env.RUN_ID }}/values.yaml"
```
It seems like this should happen somewhere deep inside install:nested:env, but not here. The abstraction has leaked.

yachmenevas replied:
Fixed

Isteb4k commented on ci/dvp-e2e/charts/cluster-config/templates/virtualization.yaml:
```diff
@@ -0,0 +1,26 @@
+---
+{{- if and (hasKey .Values "features") (.Values.features.virtualization) }}
```
What features are these? The whole point of the nightly tests is to test virtualization. It definitely needs to be enabled.

Isteb4k commented on .github/workflows/e2e-matrix.yml:
```diff
@@ -126,39 +129,17 @@ jobs:
+          task parent:kubeconfig OUTPUT="$KCFG" API_URL="${E2E_K8S_URL}" SA_TOKEN="${{ secrets.E2E_NESTED_SA_SECRET }}"
+          echo "KUBECONFIG=$KCFG" >> "$GITHUB_ENV"
 
       - name: Prepare run values.yaml
```
The abstraction has leaked. It should be inside install:nested:env

Isteb4k commented on .github/workflows/e2e-matrix.yml:
```diff
@@ -167,119 +148,94 @@ jobs:
             RUN_DIR="${{ env.TMP_ROOT }}/runs/${{ env.RUN_ID }}"
-          echo "VALUES_TEMPLATE_FILE=${{ env.TMP_ROOT }}/runs/${{ env.RUN_ID }}/values.yaml" >> $GITHUB_ENV
+
+      - name: Configure registry auth for installer pull
```
The abstraction has leaked. It should be inside install:nested:env

Isteb4k commented on .github/workflows/e2e-matrix.yml:
```diff
@@ -26,67 +26,73 @@ on:
-
     env:
-      GO_VERSION: "1.24.6"
+      PROFILE: ${{ needs.setup-nested-envs.outputs.profile }}
```
Why not to setup `RUN_ID="${{ needs.setup-nested-envs.outputs.run_id }}"` here as well?

Isteb4k commented on .github/workflows/e2e-matrix.yml:
```diff
@@ -167,119 +148,94 @@ jobs:
 
-      - name: Build parent kubeconfig from secret (cleanup)
-        shell: bash
+      - name: Build parent kubeconfig from secret
```
It should be inside cleanup:namespaces

yachmenevas committed:
refactor(ci): consolidate e2e workflow, move bootstrap to task, use matrix include

yachmenevas commented on .github/workflows/e2e-matrix.yml:
```diff
@@ -26,67 +26,73 @@ on:
+          PROFILE="${{ steps.load.outputs.profile }}"
+          echo "run_id=$RUN_ID" >> "$GITHUB_OUTPUT"
+          mkdir -p ./tmp/run-context
+          {
```
Removed

yachmenevas commented on ci/dvp-e2e/charts/cluster-config/templates/e2e-sa.yaml:
```diff
@@ -0,0 +1,27 @@
+- kind: ServiceAccount
+  name: e2e-runner
+  namespace: kube-system
+
```
Removed

yachmenevas committed:
chore(ci): read data disk count from matrix
